frontend/userpatch.lua

```diff
@@ -0,0 +1,138 @@
+    end
+    local sorting = function(a, b)
+        return tostring(a):gsub("%.?%d+", addLeadingZeroes)..("%3d"):format(#b)
+            < tostring(b):gsub("%.?%d+", addLeadingZeroes)..("%3d"):format(#a)
+    end
+
+    table.sort(patches, sorting)
+
+    for i, entry in ipairs(patches) do
+        local fullpath = dir .. "/" .. entry
+        if lfs.attributes(fullpath, "mode") == "file" then
+            if fullpath:match("%.lua$") then -- execute patch-files first
+                logger.info("Live update apply:", fullpath)
+                local ok, err = pcall(dofile, fullpath)
+                if not ok then
+                    logger.warn("Live update", err)
+                    if priority >= userpatch.late then -- Only show InfoMessage, when late during startup.
+                        -- Only developers (advanced users) will use this mechanism.
+                        -- A warning on a patch failure after an OTA update will simplify troubleshooting.
+                        local UIManager = require("ui/uimanager")
+                        local InfoMessage = require("ui/widget/infomessage")
+                        UIManager:show(InfoMessage:new{text = "Error loading patch:\n" .. fullpath}) -- no translate
+                    end
+                end
+            elseif fullpath:match("%.sh$") then -- execute shell scripts secont
+                execute("sh", fullpath, data_dir, package_dir)
+            end
+        end
+    end
+
+    -- Only delete update once marker if `early_once` updates have been applied.
+    if priority == userpatch.early_once and update_once_pending then
+        os.remove(update_once_marker) -- Prevent another execution on a further starts.
+    end
+end
+
+--- This function executes sripts and applies lua patches from `/koreader/userscripts`
+---- @string priority ... one of "early\_once", "early", "late", "before\_exit", "on\_exit"
+function userpatch.applyPatches(priority)
+    local patch_dir = data_dir .. "/userpatches"
+
+    if priority == userpatch.early then
+        -- Move an existing `koreader/patch.lua` to `koreader/userpatches/1-patch.lua` (-> will be excuted in `early`)
+        if lfs.attributes(data_dir .. "/patch.lua", "mode") == "file" then
+            if lfs.attributes(patch_dir, "mode") == nil then
+                if not lfs.mkdir(patch_dir, "mode") then
+                    logger.err("Live update error creating directory", patch_dir)
+                end
+            end
+            os.rename(data_dir .. "/patch.lua", patch_dir .. "/" .. userpatch.early .. "-patch.lua")
+        end
+    end
+
+    local update_once_marker = package_dir .. "/update_once.marker"
+    local update_once_pending = lfs.attributes(update_once_marker, "mode") == "file"
+
+    if priority >= userpatch.early or update_once_pending then
+        runLiveUpdateTasks(patch_dir, priority, update_once_pending, update_once_marker)
+    end
```

Comment:
... here, so you don't need to pass the `update_once_*` parameters and all this logic stays in these few lines here.